src/Microsoft.Net.Http.Headers/NameValueHeaderValue.cs

```diff
@@ -142,6 +143,30 @@ public override bool Equals(object obj)
             }
         }
 
+        public StringSegment GetUnescapedValue()
+        {
+            if (!HeaderUtilities.IsQuoted(_value))
+            {
+                return _value;
+            }
+            return UnescapeAsQuotedString(_value);
+        }
+
+        public void SetAndEscapeValue(StringSegment value)
+        {
+            HeaderUtilities.ThrowIfReadOnly(IsReadOnly);
+            if (StringSegment.IsNullOrEmpty(value) || (GetValueLength(value, 0) == value.Length))
+            {
+                _value = value;
+            }
+            else
+            {
+                var encodedValue = EscapeAsQuotedString(value);
+                CheckValueFormat(encodedValue);
```

Comment:
Relates to comment, the call to CheckValueFormat should probably throw (which calls HttpRuleParser) should probably throw if there is an invalid character in the string (like '\n').